### ISSUE TITLE
Causal influence agents do not learn

### ISSUE THREAD

**Issue by internetcoffeephone** [eugenevinitsky/sequential_social_dilemma_games#159]
When using the [script](https://github.com/eugenevinitsky/sequential_social_dilemma_games/blob/visible_actions/run_scripts/train_a3c_influence.py ) in the visible_actions branch, combined with the ray branch of [causal_ppo](https://github.com/natashamjaques/ray/tree/causal_ppo), the agents seem to keep hovering around 0 reward.

Some caveats: the entropy in the run script should be corrected to be negative, and the [causal_a3c](https://github.com/natashamjaques/ray/tree/causal_a3c) branch of ray is not used because it contains an error where loss_weight is not applied in [a3c_policy_graph_causal](https://github.com/natashamjaques/ray/blob/causal_a3c/python/ray/rllib/agents/a3c/a3c_policy_graph_causal.py). Additionally, KL 0/NaN handling is done differently, see [this file](https://github.com/natashamjaques/ray/blob/causal_ppo/python/ray/rllib/agents/a3c/a3c_policy_graph_causal.py). Finally, the ray 0.6.4 requirement should be removed from requirements.txt.

Can the results of the paper be reproduced like this, or am I incorrectly using the scripts?

**Comment by eugenevinitsky** [comment]
Hi, as we've warned in the setup instructions, that branch is untested. It's plausible that there are bugs there! We wrote up the env but never got around to properly testing that code out. If you do manage to get it to work please let me know! 

P.s. thank you for pointing out the entropy sign and the other mistakes in the branch.

**Comment by eugenevinitsky** [comment]
Do you have a branch up with a more correct implementation?

**Comment by internetcoffeephone** [comment]
Sure! I just made the branches, [Ray](https://github.com/internetcoffeephone/ray/tree/reproduction/causal) and [SSD](https://github.com/internetcoffeephone/sequential_social_dilemma_games/tree/reproduction/causal). These contain the fixes mentioned above, and some other fixes that stop deprecation warnings from flooding the debug log.

Note that I changed run script names to enforce more consistency. The influence moa model training is started by running train_influence_moa.py.
(I also implemented a config file, but I understand that it's hard to safely merge in with all these name changes).

**Comment by eugenevinitsky** [comment]
Thank you! I will also be testing these hopefully this week, I'll let you know if we get any good results.

**Comment by internetcoffeephone** [comment]
Please let me know if you get any bad/unexpected results too, not just good results. :)

I'd like to know in which direction I should be looking to debug the behavior. As @natashamjaques mentioned, the algorithm is very sensitive to hyperparameters. Increasing the entropy reward could help to deal with the premature convergence, but I lack the computational resources to do a large sweep over many possible combinations of learning rate/entropy/moa reward.

There could just be a bug in a3c_policy_graph_causal, but I wouldn't know where/how to start looking.

**Comment by internetcoffeephone** [comment]
Upon inspection, it appears to me that the model conv_to_fc_net_actions.py does not have the structure that is explained in the paper.
![image](https://user-images.githubusercontent.com/5096835/65983320-88e0f580-e47d-11e9-8235-421f6ec28306.png)

Instead, it looks more like this: 
![image](https://user-images.githubusercontent.com/5096835/65983548-086ec480-e47e-11e9-9a2c-d51c82a1b520.png)

The last_layer property of conv_to_fc_net_actions is fed to both LSTMs in the function get_double_lstm_model in catalog.py. last_layer has a size of (?,72): it consists of a concatted 32 (the last FC layer) + 40 (others_actions).

Am I interpreting this correctly, or does the model actually work according to the upper picture? I cannot find the separate double FC layers, colored red in the upper picture.

**Comment by eugenevinitsky** [comment]
Hi! I have computational resource limits too. The one thing I'm thinking to try to alleviate that is some of the hyperparameter tuning algorithms in Ray. You could try some of those? I will of course let you know how the project progresses, including any negative results.

My current plan is to port the code to the newest Ray version which should let me use tensorflow eager to set breakpoints in the graph and hopefully add some tests. Also, I think you are correct about the model structure. That is an error. Good catch!

**Comment by internetcoffeephone** [comment]
Cheers! I'll try to run some experiments as soon as the model has the correct structure. I've gotten started on fixing this.

Just in case you weren't aware: you can use ray.init(local_mode=True) to set breakpoints and debug the model/policy graph as well, although it's less powerful than tf eager.

I'm also wondering what the actual performance bottlenecks are - I tried profiling but couldn't make much sense of the results. Would it be nnet evaluation or stepping over the environment? Adding GPUs doesn't seem to make much of a difference, even when adding enough workers to facilitate high% GPU/CPU usage.

**Comment by internetcoffeephone** [comment]
>I've gotten started on fixing this.

Disregard this, I saw you already [got started](https://github.com/eugenevinitsky/sequential_social_dilemma_games/commit/3aa254a7ed68caff2fd791ad02944d704306c1a4). The branch seems to be missing the orientation refactor among others though, is that intentional?

**Comment by eugenevinitsky** [comment]
I'm actually not sure what the performance bottlenecks are. The missing orientation refactor is an accident, I'll merge it in shortly (it shouldn't change anything, it's just so that the env behaves more like you'd expect it to with regards to commands like up or down). It's not 100% complete yet but the current branch is almost there.

**Comment by eugenevinitsky** [comment]
If you take a look and spot any bugs let me know!

**Comment by eugenevinitsky** [comment]
@internetcoffeephone it's now in working mode. That's not to say the hyperparameters are tuned or that there aren't bugs but I think it's pretty close.

**Comment by TianChenone** [comment]
when I set the sequential social dilemma branch as visible_action or master, set the ray branch as a3c_causal. And when I run script "train_baseline.py", there are errors "No module named 'social_dilemmas'". I used to think it is caused by the path of social_dilemmas fold, so I add "import sys sys.path.append('../')" in front of import social_dilemmas, but the error still appears. I am quite confused, can you show the path of your code, or give me some suggestions? thanks very much~! ps the test results of environments are 22 passed, 135 warnings.

**Comment by natashamjaques** [comment]
FYI there is now a working implementation here: https://github.com/eugenevinitsky/sequential_social_dilemma_games/pull/179/files
